### PR TITLE
Add support for Llama 3.1

### PR DESCRIPTION
**Pull request by EricLBuehler**
Refs #2354. Llama 3.1 simply requires a modification to the way we calculate the `inv_freq`s. This PR implements it (see EricLBuehler/mistral.rs#618), and fixes some Clippy lints.

### CODE REVIEW

**Comment by BradyBonnette** [comment]
Hi @EricLBuehler 

I just tried this quickly in my environment, and I think there's a `config.json` difference that needs to be accounted for. For example, modifying `candle-examples/examples/llama/main.rs`  and changing 

```diff
@@ -132,7 +132,7 @@ fn main() -> Result<()> {
             Which::V1 => "Narsil/amall-7b".to_string(),
             Which::V2 => "meta-llama/Llama-2-7b-hf".to_string(),
             Which::V3 => "meta-llama/Meta-Llama-3-8B".to_string(),
-            Which::V3Instruct => "meta-llama/Meta-Llama-3-8B-Instruct".to_string(),
+            Which::V3Instruct => "meta-llama/Meta-Llama-3.1-8B-Instruct".to_string(),
             Which::Solar10_7B => "upstage/SOLAR-10.7B-v1.0".to_string(),
             Which::TinyLlama1_1BChat => "TinyLlama/TinyLlama-1.1B-Chat-v1.0".to_string(),
         });
```

and running `cargo run --example llama --features=cuda,cudnn,flash-attn -- --which v3-instruct --prompt "Tell me something about black holes in a single sentence."`

results in this:

```
     Running `target/debug/examples/llama --which v3-instruct --prompt 'Tell me something about black holes in a single sentence.'`
loading the model weights from meta-llama/Meta-Llama-3.1-8B-Instruct
Error: invalid type: sequence, expected u32 at line 8 column 18
```

I think it's because the `config.json` for Llama3 has the end of sentence token id (`eos_token_id`) as a single number (usize), whereas the Llama3.1 version has `eos_token_id` as an array of numbers (usize).

**Comment by EricLBuehler** [comment]
@BradyBonnette I added support for that now! Also, there is now the `--v3-1-instruct` selector.

**Comment by EricLBuehler** [comment]
@BradyBonnette were the tests successful?

**Comment by BradyBonnette** [comment]
@EricLBuehler 

The `candle-examples` tests I ran were successful. It inferences quickly and wasn't gibberish. Nice work! It kinda went off the rails a bit when I gave it an "instruction", but to be fair so did the already-established llama v3 setup that existed before your work today. I think the example app that is provided is more for text completion, but I am more focused on things like document summarization, which doesn't work too well in `candle-examples` until you provide a proper prompt.

The real test I want to perform is to integrate what you have in this branch into our project to see how it behaves. Ill report back when I have something. I dont know how quick HuggingFace is to getting PRs reviewed and merged in, I figure I have some time to test out what I need in our project.

**Comment by BradyBonnette** [comment]
@EricLBuehler 

Something is still missing. I quickly incorporated your branch into our project, and fed it a document that was 8608 tokens. It failed with this:

```
config: LlamaConfig { hidden_size: 4096, intermediate_size: 14336, vocab_size: 128256, num_hidden_layers: 32, num_attention_heads: 32, num_key_value_heads: Some(8), rms_norm_eps: 1e-5, rope_theta: 500000.0, bos_token_id: Some(128000), eos_token_id: Some(LlamaEosToks(Right([128001, 128008, 128009]))), rope_scaling: Some(Llama3RopeConfig { factor: 8.0, low_freq_factor: 1.0, high_freq_factor: 4.0, original_max_position_embeddings: 8192, rope_type: Llama3 }) }


Starting inferencing...
Got 8608 tokens
Getting logits processor
Could not inference model because: Could not forward input on Tensor because: narrow invalid args start + len > dim_len: [4096, 64], dim: 0, start: 0, len:8608
```

I suspect it has something to do with the fact that [this const](https://github.com/EricLBuehler/candle/blob/llama_31/candle-transformers/src/models/llama.rs#L7) is still being used to set the dimensions of some things, when it should be coming from the `config.json` values (I think?). I do remember this being an issue with Llama 3.0, as it's trained with a 8k window but `candle` hard codes it to 4096 since I believe that's what Llama 2 used.

**Comment by BradyBonnette** [comment]
Forgot to mention:

I don't know if this helps, but a couple weeks ago I ended up making a fork to address specifically the issue around the hard-coded context length in the llama models.

Im not sure it's 100% accurate or even right, but it worked for the purposes in which I needed it: https://github.com/huggingface/candle/compare/main...BradyBonnette:candle:brady/llama-context-size

File: .vscode/settings.json [hunk 1]
```diff
@@ -7,5 +7,5 @@
         "candle-pyo3"
     ],
     "python.testing.unittestEnabled": false,
-    "python.testing.pytestEnabled": true
+    "python.testing.pytestEnabled": true,
```
Comment (LaurentMazare) [inline .vscode/settings.json:10]: Why has this file been modified? This doesn't seem like valid json anymore or am I missing something?

File: Cargo.toml [hunk 2]
```diff
@@ -73,6 +73,7 @@ tracing-subscriber = "0.3.7"
 yoke = { version = "0.7.2", features = ["derive"] }
 zip = { version = "1.1.1", default-features = false }
 metal = { version = "0.27.0", features = ["mps"]}
+either = { version = "1.13.0", features = ["serde"] }
```
Comment (LaurentMazare) [inline Cargo.toml:76]: Let's not introduce new dependencies please. Especially here it's pretty straightforward to do without it.

File: candle-examples/Cargo.toml [hunk 3]
```diff
@@ -35,7 +35,8 @@ serde = { workspace = true }
 serde_json = { workspace = true }
 symphonia = { version = "0.5.3", features = ["all"], optional = true }
 tokenizers = { workspace = true, features = ["onig"] }
-cpal= { version = "0.15.2", optional = true }
+cpal = { version = "0.15.2", optional = true }
+either = { workspace = true }
```
Comment (LaurentMazare) [inline candle-examples/Cargo.toml:39]: Same as above.

File: candle-transformers/Cargo.toml [hunk 4]
```diff
@@ -15,6 +15,7 @@ byteorder = { workspace = true }
 candle = { workspace = true }
 candle-flash-attn = { workspace = true, optional = true }
 candle-nn = { workspace = true }
+either = { workspace = true }
```
Comment (LaurentMazare) [inline candle-transformers/Cargo.toml:18]: Same as above.

File: .vscode/settings.json (hunk 1 above)
Comment (EricLBuehler) [reply .vscode/settings.json]: Fixed that, my mistake!

File: Cargo.toml (hunk 2 above)
Comment (EricLBuehler) [reply Cargo.toml]: I removed the either dep throughout.

**Comment by EricLBuehler** [comment]
@BradyBonnette I updated it so that the max position embeddings are now taken from the config.

**Review by LaurentMazare — APPROVED** [review]

**Comment by EricLBuehler** [comment]
Thank you!

**Comment by BradyBonnette** [comment]
@EricLBuehler I had to step away for a bit, but that's fantastic! Thanks a ton for doing this. Got merged before I could do another test, but I'm gonna do that shortly regardless :)

**Comment by EricLBuehler** [comment]
Sounds good, please let me know what the result is!

**Comment by BradyBonnette** [comment]
@EricLBuehler 

Ive been testing this out extensively in our project, and it's been going great. It seems inference small documents and large documents just fine.

Thank you for jumping on it so quick and getting something workable in short order. Much appreciated!

**Pull request merged**